File: src/main/java/com/nurkiewicz/tsclass/parser/visitors/ExpressionVisitor.java
package com.nurkiewicz.tsclass.parser.visitors;

import com.nurkiewicz.tsclass.antlr.parser.TypeScriptBaseVisitor;
import com.nurkiewicz.tsclass.antlr.parser.TypeScriptParser;
import com.nurkiewicz.tsclass.parser.ast.expr.Expression;

class ExpressionVisitor extends TypeScriptBaseVisitor<Expression> {

        @Override
        public Expression visitAssignmentExpression(TypeScriptParser.AssignmentExpressionContext ctx) {
            return ctx.accept(new AdditiveExpressionVisitor());
        }

}

File: src/main/kotlin/com/nurkiewicz/tsclass/parser/visitors/ExpressionVisitor.kt
package com.nurkiewicz.tsclass.parser.visitors

import com.nurkiewicz.tsclass.antlr.parser.TypeScriptBaseVisitor
import com.nurkiewicz.tsclass.antlr.parser.TypeScriptParser
import com.nurkiewicz.tsclass.parser.ast.Expression
import com.nurkiewicz.tsclass.parser.ast.Relational

internal object ExpressionVisitor : TypeScriptBaseVisitor<Expression>() {

    override fun visitRelationalExpression(ctx: TypeScriptParser.RelationalExpressionContext): Expression {
        return if (ctx.relationalOperator() != null) {
            val left = ctx.relationalExpression().accept(this)
            val right = ctx.shiftExpression().accept(AdditiveExpressionVisitor())
            val operator = Relational.Operator.of(ctx.relationalOperator().text)
            Relational(left, operator, right)
        } else {
            ctx.accept(AdditiveExpressionVisitor())
        }
    }
}